Minio/ApiEndpoints/ObjectOperations.cs

```diff
@@ -1609,26 +1614,32 @@ private async Task<object> CopyObjectRequestAsync(CopyObjectRequestArgs args, Ca
     /// <returns>bytes read in a byte array</returns>
     internal async Task<byte[]> ReadFullAsync(Stream data, int currentPartSize)
     {
-        var result = new byte[currentPartSize];
         var totalRead = 0;
-        while (totalRead < currentPartSize)
-        {
-            var curData = new byte[currentPartSize - totalRead];
-            var curRead = await data.ReadAsync(curData, 0, currentPartSize - totalRead).ConfigureAwait(false);
-            if (curRead == 0) break;
-            for (var i = 0; i < curRead; i++) result[totalRead + i] = curData[i];
-            totalRead += curRead;
-        }
+        var shared = ArrayPool<byte>.Shared;
+        var curData = shared.Rent(currentPartSize - totalRead);
 
-        if (totalRead == 0) return null;
+        try
+        {
+            while (totalRead != currentPartSize &&
+                   totalRead < currentPartSize)
+            {
+                curData = shared.Rent(currentPartSize - totalRead);
+                var curRead = await data.ReadAsync(curData, 0, currentPartSize - totalRead).ConfigureAwait(false);
+                if (curRead == 0) break;
+                totalRead += curRead;
+            }
 
-        if (totalRead == currentPartSize) return result;
+            if (totalRead == 0) return null;
 
-        var truncatedResult = new byte[totalRead];
-        for (var i = 0; i < totalRead; i++) truncatedResult[i] = result[i];
-        return truncatedResult;
+            return curData;
```

Comment:
You can't return a rented buffer that is returned the the pool below on line 1638.

Reply:
Yes.
This is where I have a `try{} / finally{}` structure and a `shared.Return` in the coming commit.